tests/Spec.hs

```diff
@@ -145,142 +149,160 @@ defaultResults = Results
 
-takeResultGetPullRequest :: (HasCallStack, Monoid w) => RWS r w Results (Maybe GithubApi.PullRequest)
+takeResultGetPullRequest :: (HasCallStack, State Results :> es) => Eff es (Maybe GithubApi.PullRequest)
 takeResultGetPullRequest =
   takeFromList
     "resultGetPullRequest"
     resultGetPullRequest
     (\v res -> res { resultGetPullRequest = v })
 
-takeResultGetOpenPullRequests :: (HasCallStack, Monoid w) => RWS r w Results (Maybe IntSet)
+takeResultGetOpenPullRequests :: (HasCallStack, State Results :> es) => Eff es (Maybe IntSet)
 takeResultGetOpenPullRequests =
   takeFromList
     "resultGetOpenPullRequests"
     resultGetOpenPullRequests
     (\v res -> res { resultGetOpenPullRequests = v })
 
-takeResultGetLatestVersion :: (HasCallStack, Monoid w) => RWS r w Results (Either TagName Integer)
+takeResultGetLatestVersion :: (HasCallStack, State Results :> es) => Eff es (Either TagName Integer)
 takeResultGetLatestVersion =
   takeFromList
     "resultGetLatestVersion"
     resultGetLatestVersion
     (\v res -> res { resultGetLatestVersion = v })
 
-takeResultGetChangelog :: (HasCallStack, Monoid w) => RWS r w Results (Maybe Text)
+takeResultGetChangelog :: (HasCallStack, State Results :> es) => Eff es (Maybe Text)
 takeResultGetChangelog =
   takeFromList
     "resultGetChangeLog"
     resultGetChangelog
     (\v res -> res { resultGetChangelog = v })
 
-takeResultGetDateTime :: (HasCallStack, Monoid w) => RWS r w Results (T.UTCTime )
+takeResultGetDateTime :: (HasCallStack, State Results :> es) => Eff es T.UTCTime
 takeResultGetDateTime =
   takeFromList
     "resultGetDateTime"
     resultGetDateTime
     (\v res -> res { resultGetDateTime = v })
 
-runRetrieveInfoRws :: HasCallStack => Config.ProjectConfiguration -> RetrieveEnvironmentFree a -> RWS () [ActionFlat] Results a
-runRetrieveInfoRws projectConfig = \case
-  GetProjectConfig cont -> pure $ cont projectConfig
-  GetDateTime cont -> cont <$> takeResultGetDateTime
-  GetBaseBranch cont -> pure $ cont $ BaseBranch $ Config.branch testProjectConfig
+runRetrieveInfo
+  :: State Results :> es
+  => Config.ProjectConfiguration
+  -> Eff (RetrieveEnvironment : es) a
+  -> Eff es a
+runRetrieveInfo projectConfig = interpret $ \_ -> \case
+  GetProjectConfig -> pure projectConfig
+  GetDateTime -> takeResultGetDateTime
+  GetBaseBranch -> pure $ BaseBranch $ Config.branch testProjectConfig
+
+type ActionResults = [BaseAction, RetrieveEnvironment, State Results, Writer [ActionFlat]]
 
 -- This function simulates running the actions, and returns the final state,
 -- together with a list of all actions that would have been performed. Some
 -- actions require input from the outside world. Simulating these actions will
 -- consume one entry from the `Results` in the state.
-runBaseActionRws :: HasCallStack => BaseActionFree a -> RWS () [ActionFlat] Results a
-runBaseActionRws =
+runBaseActionResults
+  :: (State Results :> es, Writer [ActionFlat] :> es)
```

Comment:
One big warning (and generally known no-no), is that you should be very careful with writer monads, specifically with lists. The `effectful` documentation [warns](https://hackage.haskell.org/package/effectful-core-2.2.2.2/docs/Effectful-Writer-Static-Local.html) for this as well. I decided not to rewrite this, because the original code already used a writer, and I wanted to avoid refactoring beyond the rewrite itself.